### PR TITLE
etcdmain: do not use defer in for loop

### PR DESCRIPTION
**Pull request by mqliang**
remove defer statements in for loop, it's hard to read and using defer in for loop may lead source leak

### CODE REVIEW

File: etcdmain/etcd.go [hunk 1]
```diff
@@ -210,20 +210,17 @@ func startEtcd(cfg *config) (<-chan struct{}, error) {
...
 		}
+
+		urlStr := u.String()
+		plog.Info("listening for peers on ", urlStr)
+
 		var l net.Listener
 		l, err = rafthttp.NewListener(u, cfg.peerTLSInfo)
 		if err != nil {
+			l.Close()
+			plog.Info("stopping listening for peers on ", urlStr)
```
Comment (gyuho) [inline etcdmain/etcd.go]: It `NewListener` return err, the listener is `nil`, so no need to close here.
So it seems we don't even need to check error for this loop and ok to let defer run
after outer function exit.

Comment (mqliang) [inline etcdmain/etcd.go]: How about just check err for log writing and remove defer()? defer in a for loop is al little confusing.

**Comment by xiang90** [comment]
We want to close the listener when there is an error returned. This commit changes the behavior to close the listener when etcd fails to listen. This does not make sense to me.

**Comment by mqliang** [comment]
ok, close this.

**Comment by xiang90** [comment]
@mqliang Thanks for your effort!

**Pull request closed without merging**